slam_toolbox/lib/karto_sdk/src/Mapper.cpp

```diff
@@ -1605,7 +1605,11 @@ namespace karto
     // only attach link information if the edge is new
     if (isNewEdge == true)
     {
-      pEdge->SetLabel(new LinkInfo(pFromScan->GetSensorPose(), rMean, rCovariance));
+      auto corrected_pose=pToScan->GetCorrectedPose();
```

Comment:
Please explain the steps here and what's changing - I'm not able to easily follow the logic change here, but my thoughts are below. Also, why not make these changes somewhere higher up in the software, like closer to the process functions? 

AddVertex adds just the entire `pScan` object, so there's no caring about the specific frame of the data. Then the major place its used is in the optimizer (https://github.com/SteveMacenski/slam_toolbox/blob/foxy-devel/solvers/ceres_solver.cpp#L270) which uses the `GetCorrectedPose`'s to operate - so corrected poses.

Then looking at the optimizer's AddConstraints, we see we get the information from https://github.com/SteveMacenski/slam_toolbox/blob/foxy-devel/solvers/ceres_solver.cpp#L309 `GetPoseDifference` which returns the output of `Update` called in the `LinkInfo` constructor using the L1612 inputs. So those inputs should be in the same frame as `AddNode` one way or another. 

So, ok, I can understand then that we either need to change the `AddNode` in the optimizer to use the sensor poses **or** change the edges to use the corrected poses like we give to the `AddNode`. Did you try the first? That would be less changes and I _believe_ it would have the same outcome (though the pose-graph means something slightly different then). 

---

So I'm on the same page about the core problem, though the code you have here is a little confusing to me.

Reply:
Yes I have tried that "1 line change" which sets the sensor pose as the vertex node. Everything worked OK until once I tried to deserialize a posegraph file that I have just saved with the save version code. The program threw an exception and died. Maybe it assumes that all the vertices stand for the base poses (I'm not sure about this). So I tried the second way which you can see now.

Reply:
Ok - can you still walk me through these changes then? I'm not following the logic intuitively about what each of these do and why.

Reply:
My goal is to set constraints between 2 base poses instead of sensor poses. Now we need to create a `LinkInfo` object as a label of the edge to link these 2 base poses (L1612). First argument is base pose of `pFromScan` which is `pFromScan->GetCorrectedPose()`. Second argument is not simply base pose of `pToScan` but the `rMean`'s corresponding base pose. As you can see from `MapperGraph::AddEdges` function, `rMean` is a sensor pose. L1609 and L1610 calculate that base pose we need which is passed to `LinkInfo` below. As we have used `pToScan` as a tool to make a transformation (L1609), it changes the original base pose of `pToScan` (in function `LocalizedRangeScan::SetSensorPose` we called in L1609) which I'm not sure will influence the processes below or not. So I decide to reset the original base pose to `pToScan` (L1611) so that every value is the same as the original code except the arguments of `LinkInfo`.
After ceres solver optimizing the pose graph, we need to update the `LocalizedRangeScan`. The optimization result is base poses, so we call `LocalizedRangeScan::SetCorrectedPose` instead of `LocalizedRangeScan::SetSensorPose` (L1995).

Reply:
> L1609 and L1610 calculate that base pose we need which is passed to LinkInfo below. 

I totally understand that, but I'm a little concerned by the fact that you're `setting` the pose, overriding whatever was there beforehand to compute this value. Would it not be possible to just [transform the data](https://github.com/SteveMacenski/slam_toolbox/blob/b0800ed7e5b1f955386a46a0ac8f9ef159120c13/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h#L5637-L5645) in the same way shown in that function without `Setting` / `Getting`? Or do you really need to overrwrite whatever was there beforehand? I know that was easy to do for prototyping, but I think we need to justify the ramifications that will have if we update the sensor-poses of the data based on when this will be called (consecutive matching, loop closure, running scans). I think this will degrade performance in all non-optimizer iterations.  

Recommendation: Make [this logic](https://github.com/SteveMacenski/slam_toolbox/blob/b0800ed7e5b1f955386a46a0ac8f9ef159120c13/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h#L5637-L5646) into a function (or maybe an analog function to `GetSensorAt`, `GetCorrectedAt` will work?) that `SetSensorPose` uses. Then make that function public so that in L1609/1610 you can call that to replace the `set` and `get`. 

---

> So I decide to reset the original base pose to pToScan (L1611) so that every value is the same as the original code except the arguments of LinkInfo.

If you do the recommendation above, then this won't be necessary because you didn't change any of the `pToScan` data. Also its alot more efficient because the `SetSensorPose` calls `Update` which iterates through all the points. 

---

L1995 we're on the same page about, no issue. Really all `SetSensorPose` is doing anyhow is computing the offset value of `m_CorrectedPose` so you're just cutting out the middleman.

Reply:
I understand what you mean. I'll push a new commit soon.

Reply:
> > L1609 and L1610 calculate that base pose we need which is passed to LinkInfo below.
> 
> I totally understand that, but I'm a little concerned by the fact that you're `setting` the pose, overriding whatever was there beforehand to compute this value (though 1611 resets). Would it not be possible to just [transform the data](https://github.com/SteveMacenski/slam_toolbox/blob/b0800ed7e5b1f955386a46a0ac8f9ef159120c13/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h#L5637-L5645) in the same way shown in that function without `Setting` / `Getting`?
> 
> Recommendation: Make [this logic](https://github.com/SteveMacenski/slam_toolbox/blob/b0800ed7e5b1f955386a46a0ac8f9ef159120c13/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h#L5637-L5646) into a function (or maybe an analog function to `GetSensorAt`, `GetCorrectedAt` will work?) that `SetSensorPose` uses. Then make that function public so that in L1609/1610 you can call that to replace the `set` and `get`. Then no data is messed with and we just transform some data.

I have pushed a new commit. Simplify setting `LinkInfo`.